torchtune/generation/_generation.py

```diff
@@ -160,31 +318,46 @@ def generate(
         # if incremental decoding is enabled, we can use the current position
         # otherwise, we take the whole sequence up to the current position
         if incremental_decoding:
-            curr_input_pos = input_pos[curr_pos].unsqueeze(0)
+            curr_input_pos = input_pos[:, curr_pos]
+            curr_cache_pos = (
+                cache_pos[curr_pos].unsqueeze(0) if cache_pos is not None else None
+            )
+            curr_masks = masks[:, curr_pos, None, :]
         else:
-            curr_input_pos = input_pos[: curr_pos + 1]
             tokens = generated_tokens.clone()
+            curr_input_pos = input_pos[:, : curr_pos + 1]
+            curr_masks = masks[:, : curr_pos + 1, : curr_pos + 1]
 
-        tokens = custom_generate_next_token(
+        q = torch.empty(
+            (bsz, model.tok_embeddings.num_embeddings), device=prompt.device
+        ).exponential_(1, generator=rng)
```

Comment:
I feel like we've now just given a couple equivalent default definitions of this in different places which adds to the burden of understanding the code (but lmk if I'm missing something important here)

Reply:
The main reason we have to define it here is to make generate_next_token compile-friendly when using rng, but also allow generate_next_token to work independently from this kind-of-hacky logic i.e. you can do all this without having to sample outside of generate_next_token. 

I agree it makes things confusing and we could drop the default defn inside generate_next_token so it's a required arg.

Reply:
Another dumb q - why does it need to be defined every time within this loop?

Reply:
Similar to how you'd re-generate it every time in `sample` you have a new random distribution to sample from, right?